### ISSUE TITLE
Privacy Badger should support Safari

### ISSUE THREAD

**Issue by cypherpunk** [EFForg/privacybadger#549]

**Comment by cooperq** [comment]
I will gladly accept pull requests for this, or will provide assistance to a project that is doing this, but I don't have the time to make those versions.

**Comment by ecliptik** [comment]
@cooperq with the recent passage of SJ Res 34, could this be re-opened to support more browsers?

**Comment by freediverx** [comment]
Please support Safari. Firefox is an unstable mess of a browser.

**Comment by ghostwords** [comment]
It doesn't seem like Safari provides a way to intercept and potentially modify headers, something that is covered by [`chrome.webRequest`](https://developer.chrome.com/extensions/webRequest) in Chrome (and now Firefox via WebExtensions). The problem is that [`canLoad`](https://developer.apple.com/reference/safariextensions/safaricontentbrowsertabproxy/1635506-canload) offers only a subset of chrome.webRequest functionality.

So it appears that Safari is missing some functionality required by Privacy Badger to function, but this needs more investigation. Does the newer declarative content Blocking API help?

**Comment by jessesquires** [comment]
@ghostwords Is this the only blocker for Safari support, or are there others?

**Comment by ghostwords** [comment]
There are these technical issues (but they need further investigation to see whether they really are blockers or not), but then there are also legal issues with things Apple asks you to agree with if you want to publish (?) a Safari extension. The legal issues might be the real blockers. I need to investigate what those are and whether we could do anything about them.

**Comment by jessesquires** [comment]
A new notes/resources on content blocking in Safari:

- https://webkit.org/blog/3476/content-blockers-first-look/
- https://developer.apple.com/videos/play/wwdc2015/511/
- https://github.com/Erulezz/Ultimate-AdBlock

Regarding "publishing" -- the only thing here is if you want your extension listed in Apple's ["extension gallery"](https://safari-extensions.apple.com), then Apple has to approve it. I'm not sure this is absolutely necessary. EFF could just link to the GitHub latest release.

**Comment by MmmmJoel** [comment]
@jessesquires is correct. You do not need to agree to Apple's terms and pay $100 to release a Safari extension. It is only required if you want to be listed in the extension gallery.

**Comment by svandragt** [comment]
https://stackoverflow.com/a/28808411 should help with the porting effort

**Comment by cowlicks** [comment]
I doubt this would be very hard! If anyone who uses Safari (I don't own a mac) want's to port it, I'd be happy to guide you and answer any questions you have.

A warning to everyone, my first major contribution to PB was porting it to FF, and look where it got me.

**Comment by JamesMcMahon** [comment]
This may no longer be needed with High Sierra's tracking blocking in Safari. I hope the EFF does a review of the functionality when it's released to make sure it adequate for privacy minded users.

**Comment by JamesMcMahon** [comment]
They read my mind, https://www.eff.org/deeplinks/2017/09/apple-does-right-users-wrong-advertisers.

I take this as an endorsement of Apple’s implementation.

**Comment by chris1000** [comment]
I just wanted to see if this is still active and if there is anything I can do to help with the project

**Comment by ghostwords** [comment]
The main problem with offering a version of Privacy Badger for Safari is that, after legal review, EFF found Apple's developer agreement unacceptable for some of the reasons stated in [this EFF blog post](https://www.eff.org/deeplinks/2014/12/sorry-iphone-users-apples-dev-agreement-means-no-eff-mobile-app-iphone) (the post talks about iOS, but the problem also applies to Safari/macOS). I believe this problem goes beyond being listed in the extensions gallery.

Furthermore, Safari seems to lack certain extension capabilities required by Privacy Badger to function properly. Privacy Badger is not an ad blocker; Privacy Badger is different from most other privacy tools as it does not come with a list of domains to block. Instead, Privacy Badger watches and learns which third-party domains appear to track you as you browse the Web, something that doesn't seem fully supported by Safari.

I personally recommend installing uBlock Origin, an excellent privacy tool.

**Comment by ghostwords** [comment]
See also this HTTPS Everywhere thread: https://github.com/EFForg/https-everywhere/issues/5121#issuecomment-343277620

**Comment by fuzzy76** [comment]
Can we get a new look on this, since Safari in the new macOS will support webextensions?

**Comment by ghostwords** [comment]
The biggest obstacle to supporting Safari at this time is that as per [Apple documentation](https://developer.apple.com/documentation/safariservices/safari_web_extensions/assessing_your_safari_web_extension_s_browser_compatibility), `webRequest` does not support blocking or modifying requests/responses. As [AdGuard write on their blog](https://adguard.com/en/blog/safari-web-extensions-response.html), Safari's version of the webRequest API is "useless for ad blocker developers". (Privacy Badger is not an ad blocker but has similar API requirements.)

This doesn't mean we won't have a version of Privacy Badger in the future. For example, we could build a version of Privacy Badger as a [Safari content blocker app](https://developer.apple.com/documentation/safariservices/creating_a_content_blocker). However, as this will take significant effort to build and maintain, it competes for attention with similar but more urgent efforts like a Manifest V3 version of Privacy Badger for Chrome (#2273).

**Comment by 10Meisterbaelle** [comment]
> Please support Safari. Firefox is an unstable mess of a browser.

You can get [ungoogled-chromium](https://github.com/Eloston/ungoogled-chromium)😉

**Comment by webjive** [comment]
Now that Apple has announced that they want to help unify extensions across browsers, maybe they'll get with the program. Not being able to easily port extensions between browsers is what's holding Safari back, to this day.

**Comment by adriank1410** [comment]
Has the situation changed now?

**Comment by ghostwords** [comment]
No updates since my [comment above](https://github.com/EFForg/privacybadger/issues/549#issuecomment-744583479).

> webRequest
> Not supported in iOS.
> 
> BlockingResponse not supported.
> 
> **Blocking requests not supported.**
> 
> opt_extraInfoSpec not supported for any of the events.

-- https://developer.apple.com/documentation/safariservices/safari_web_extensions/assessing_your_safari_web_extension_s_browser_compatibility

**Comment by dijonkitchen** [comment]
Isn't it just not supported on iOS? Can it be ported to macOS Safari?

**Comment by ghostwords** [comment]
I think macOS Safari and iOS Safari use the same base extensions system, with iOS having some additional restrictions/considerations. For example, all Safari extensions lack blocking capability using `webRequest`, as per https://github.com/EFForg/privacybadger/issues/549#issuecomment-962673636, and `webRequest` is missing entirely on iOS. My previous comment applies to all Safari extensions.

Having said that, we do want to bring Privacy Badger to Safari. I don't know when it will happen.

**Comment by leonpano2006** [comment]
safari already do the job of privacy badger and that is built in

**Comment by earthwormjim2** [comment]
> safari already do the job of privacy badger and that is built in

Are you referring to this Safari feature?: [Safari Tracking Prevention](https://www.howtogeek.com/327773/how-safaris-new-intelligent-tracking-prevention-works/)

**Comment by leonpano2006** [comment]
> > safari already do the job of privacy badger and that is built in
> 
> Are you referring to this Safari feature?: [Safari Tracking Prevention](https://www.howtogeek.com/327773/how-safaris-new-intelligent-tracking-prevention-works/)

yes
it basically do what Privacy Badger does

**Comment by JW-Moose** [comment]
> > > safari already do the job of privacy badger and that is built in
> > 
> > 
> > Are you referring to this Safari feature?: [Safari Tracking Prevention](https://www.howtogeek.com/327773/how-safaris-new-intelligent-tracking-prevention-works/)
> 
> yes it basically do what Privacy Badger does

I tend to use multiple tools that provide this functionality, and typically one tool doesn't catch them all. For example, on Firefox, I use DuckDuckGo, Privacy Badger, and Firefox's built-in tracking protection tool, but PB and DDG add-ons regularly block additional things. 

Defense in depth applies to privacy too for sure :) I'd love to see Privacy Badger on Safari.

**Comment by jwilker** [comment]
Plus 1 to @JW-Moose. I use lots of tools on macOS, so relying on just one for iOS wouldn't be my preference, but certainly, it's better than none. I'd still love PB on iOS/iPadOS

**Comment by Cervantes12** [comment]
> 

Update
My  update

**Comment by leonpano2006** [comment]
Wtf is this minimized comment

**Comment by ghostwords** [comment]
To follow up on https://github.com/EFForg/privacybadger/issues/549#issuecomment-1079096990, we may be able to offer Privacy Badger for desktop/macOS (not mobile/iOS) Safari in the first half of next year.

**Comment by enr00ted** [comment]
Any news on this ? :)

**Comment by BitBackchat** [comment]
MacOS comes with a built in track blocker. How does that & other proprietary tracker blockers compare to PrivacyBadger?

**Comment by ghostwords** [comment]
> MacOS comes with a built in track blocker. How does that & other proprietary tracker blockers compare to Privacy Badger?

Safari's built-in [Tracking Protection does not actually block](https://www.simoahava.com/analytics/no-safari-does-not-block-google-analytics/) trackers from loading. It's better than nothing (Chrome), but worse than having a real tracker blocker.

>other proprietary tracker blockers

See https://privacybadger.org/#Is-Privacy-Badger-compatible-with-other-extensions%2c-including-other-adblockers.

**Comment by OIRNOIR** [comment]
> > MacOS comes with a built in track blocker. How does that & other proprietary tracker blockers compare to Privacy Badger?
> 
> Safari's built-in [Intelligent Tracking Prevention does not actually block trackers](https://www.simoahava.com/analytics/no-safari-does-not-block-google-analytics/) from loading. It's better than nothing (Chrome), but worse than having a real tracker blocker.
> 
> > other proprietary tracker blockers
> 
> See https://privacybadger.org/#Is-Privacy-Badger-compatible-with-other-extensions%2c-including-other-adblockers.

Is this still true with MacOS 14 (Sonoma)'s new revision of ITP?
<img width="392" alt="Screenshot 2023-10-07 at 10 17 51 PM" src="https://github.com/EFForg/privacybadger/assets/70721372/552dcc55-983d-4598-b473-c53518a37e59">

**Comment by ghostwords** [comment]
Safari 17 [blocks "known trackers and fingerprinting" in Private Browsing windows](https://developer.apple.com/documentation/safari-release-notes/safari-17-release-notes#Private-Browsing).

Why only in Private Browsing? My guess is because full tracker blocking can and will break websites.

If you only browse in Private Browsing windows and if you are happy with how much tracking is blocked and how easy it is to disable blocking selectively when something breaks, then you should be all set.

**Comment by OIRNOIR** [comment]
> Safari 17 [blocks "known trackers and fingerprinting" in Private Browsing windows](https://developer.apple.com/documentation/safari-release-notes/safari-17-release-notes#Private-Browsing).
> 
> Why only in Private Browsing? My guess is because full tracker blocking can and will break websites.
> 
> If you only browse in Private Browsing windows and if you are happy with how much tracking is blocked and how easy it is to disable blocking selectively when something breaks, then you should be all set.

Just FYI, you can enable it for all browsing instead of private browsing.

<img width="629" alt="Screenshot 2023-10-08 at 11 40 04 AM" src="https://github.com/EFForg/privacybadger/assets/70721372/ed68f0c2-d9e6-419b-8cae-434a759fdc45">

Thanks for the information, though!

**Comment by nirnaeth-arnoediad** [comment]
What about Firefox on iOS? May we activate somehow?

**Comment by ghostwords** [comment]
> What about Firefox on iOS? May we activate somehow?

Firefox on iOS [does not support extensions](https://mozilla.github.io/platform-tilt/#issue_15).

This issue is about releasing Privacy Badger for Safari. If your question is not related to Safari support, please start a new issue or discussion instead of commenting here.

**Comment by misanthroope** [comment]
Privacy Badger should support Safari in 2024

**Comment by iretrala** [comment]
Looking forward to this. Any idea what quarter it might be released?

**Comment by ghostwords** [comment]
>Looking forward to this. Any idea what quarter it might be released?

Not too long after the Manifest V3 version of Privacy Badger for Chrome, which as of today needs to happen before June this year.

**Comment by misanthroope** [comment]
> Privacy Badger should support Safari in 2024

You can do it, we're rooting for you!

**Comment by KinetekEnergy** [comment]
any new updates on this? :D

**Comment by ghostwords** [comment]
> any new updates on this? :D

We recently finished updating Privacy Badger to Manifest V3 in Chrome. We are now beginning to look into Privacy Badger for macOS, using the Chrome MV3 version as the starting point. I expect we will find new bugs and missing pieces of WebExtensions functionality, which we will report to Apple.

**Comment by misanthroope** [comment]
> > Privacy Badger should support Safari in 2024
> 
> You can do it, we're rooting for you!

before 2024, good luck!!

**Comment by RareMinnow** [comment]
@ghostwords any status updates on this?

**Comment by lenacohen** [comment]
We're actively working on bringing Privacy Badger to Safari, but the process is not straightforward due to [limitations](https://github.com/duckduckgo/mv3-compat-tests?tab=readme-ov-file#results) in Safari's web extension APIs. For instance, we may no longer be able to list blocked domains in the Privacy Badger popup because Safari’s webRequest API doesn’t fire events for requests blocked by DNR rules. While declarativeNetRequest.getMatchedRules is a potential workaround, it is not consistently reliable on Safari. Another major issue is Safari’s lack of support for DNR rule conditions like excludedInitiatorDomains, which Privacy Badger uses to prevent site breakages by not blocking requests from [multi-domain first parties](https://github.com/EFForg/privacybadger/blob/master/src/js/multiDomainFirstParties.js). Support for excludedInitiatorDomains appears to be [in development](https://developer.apple.com/documentation/safari-technology-preview-release-notes/stp-release-208#New-Features) for Safari. We’re making slow and steady progress working around these limitations!

<img width="777" alt="Image" src="https://github.com/user-attachments/assets/966e7db2-0ced-40f7-ba16-248b1f838cc7" />

**Comment by Tarknassus** [comment]
Just to say I'd love to see this on Safari - Apple's own tracking and fingerprinting service is pretty vague and fairly binary.  By which I mean services like Cloudflare's 'Are You Human' will NOT work at all when the service is on, with zero ways to allow enough to get it to function.  You have to turn the whole thing off.

Privacy Badger's simple UI would be awesome.  Happy to beta test too ;)

**Comment by dertuxmalwieder** [comment]
"Just use a different browser" is a horrible advice.

**Comment by soklovsk** [comment]
How is it going?

**Comment by ghostwords** [comment]
We are blocked by several bugs in Safari: https://github.com/EFForg/SafariWebExtensionBugs/issues?q=is%3Aissue%20state%3Aopen%20label%3Acritical

**Comment by Squuiid** [comment]
> We are blocked by several bugs in Safari: https://github.com/EFForg/SafariWebExtensionBugs/issues?q=is%3Aissue%20state%3Aopen%20label%3Acritical

These still an issue with the release of Tahoe?

**Comment by Julmik647** [comment]
You said about some limitations of Safari iOS, so on iOS you can make Privacy Badger browser. Yes, I know that it still needs to be webkit, but it may have less limitations than Safari iOS extension.

> Privacy Badger should support Safari in 2024

It's 2025 and it still doesn't support.

**Comment by ghostwords** [comment]
> > We are blocked by several bugs in Safari: https://github.com/EFForg/SafariWebExtensionBugs/issues?q=is%3Aissue%20state%3Aopen%20label%3Acritical
> 
> These still an issue with the release of Tahoe?

We'll retest and update our [Safari bugs repository](https://github.com/EFForg/SafariWebExtensionBugs/) in the next few weeks.

**Comment by ghostwords** [comment]
If you'd like to help us test Privacy Badger for Safari on macOS, you can now join our [TestFlight](https://testflight.apple.com/) using the following public link:

https://testflight.apple.com/join/M12n4GyM

Any data collected by EFF as part of this project is covered by EFF’s privacy policy: https://www.eff.org/policy. Data collected by Apple or any other third parties is covered by their respective privacy policies.
